### PR TITLE
SSH request tracing

### PR DESCRIPTION
**Pull request by rosstimothy**
Add tracing support for ssh global requests and  channels. Wrappers
for `ssh.Client`, `ssh.Channel`, and `ssh.NewChannel` provide a
mechanism for tracing context to be propagated via a `context.Context`.

In order to maintain backwards compatibility the ssh.Client wrapper
tries to open a TracingChannel when constructed. Any servers that
don't support tracing will reject the unknown channel. The client
will only provide tracing context to servers which do NOT reject
the TracingChannel request.

In order to include pass tracing context along all ssh payloads
are wrapped in an Envelope that includes the original payload
AND any tracing context. Servers now try to unmarshal all payloads
into said Envelope when processing messages. If there is an Envelope
provided, a new span will be created and the original payload will
be pass along to handlers.

Part of #12241

### CODE REVIEW

File: api/observability/tracing/ssh/ssh_test.go [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+	})
+	require.NoError(t, err)
+
+	return sconn, nc, r
+}
+
+func newServer(t *testing.T, handler func(*ssh.ServerConn, <-chan ssh.NewChannel, <-chan *ssh.Request)) *server {
+	listener, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+
```
Comment (zmb3) [inline api/observability/tracing/ssh/ssh_test.go:107]: Do we need a t.Cleanup(func() { listener.Close() })` here?

File: api/observability/tracing/ssh/client_test.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+						_, _, err := ch.Accept()
+						if err != nil {
+							errChan <- trace.Wrap(err, "failed to accept channel")
+							return
+						}
+					}
+				}
+			})
+
+			go srv.Run(errChan)
```
Comment (zmb3) [inline api/observability/tracing/ssh/client_test.go:110]: How does this goroutine terminate?

**Review by zmb3 — APPROVED** [review]

File: api/observability/tracing/ssh/ssh_test.go [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+				conn,
+				chans,
+				reqs,
+				tracing.WithTracerProvider(tp),
+				tracing.WithTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{})),
+			)
+			require.Equal(t, handler.tracingSupported, client.capability)
+
+			ctx, span := tp.Tracer("test").Start(context.Background(), "test")
+			t.Cleanup(func() { span.End() })
```
Comment (gabrielcorado) [inline api/observability/tracing/ssh/ssh_test.go:312]: Is this cleanup necessary? We're ending the `span` right after the `SendRequest` call.

File: lib/srv/forward/sshserver.go [hunk 4]
```diff
@@ -682,6 +726,22 @@ func (s *Server) handleChannel(ctx context.Context, nch ssh.NewChannel) {
 	channelType := nch.ChannelType()
 
 	switch channelType {
+	// Channels of type "tracing-request" are sent to determine if ssh tracing envelopes
+	// are supported. Accepting the channel indicates to clients that they may wrap their
+	// ssh payload with tracing context.
+	case tracessh.TracingChannel:
+		ch, _, err := nch.Accept()
+		if err != nil {
```
Comment (gabrielcorado) [inline lib/srv/forward/sshserver.go:734]: Can we log this error [like in the other channel](https://github.com/gravitational/teleport/blob/217f7240ad966b17abb9e041d4be5b9b6673998c/lib/srv/forward/sshserver.go#L903) types?

File: lib/sshutils/server.go [hunk 5]
```diff
@@ -497,16 +518,67 @@ func (s *Server) HandleConnection(conn net.Conn) {
...
 				connClosed()
 				return
 			}
-			go s.newChanHandler.HandleNewChan(ctx, ccx, nch)
+
+			// This is a request from clients to determine if tracing is enabled.
+			// Handle here so that we always alert clients that we can handle tracing envelopes.
+			if nch.ChannelType() == tracessh.TracingChannel {
+				ch, _, err := nch.Accept()
+				if err != nil {
```
Comment (gabrielcorado) [inline lib/sshutils/server.go:553]: Can we log this error not to lose it?

File: lib/srv/regular/sshserver.go [hunk 6]
```diff
@@ -1168,6 +1185,21 @@ func (s *Server) HandleNewChan(ctx context.Context, ccx *sshutils.ConnectionCont
 	channelType := nch.ChannelType()
 	if s.proxyMode {
 		switch channelType {
+		// Channels of type "tracing-request" are sent to determine if ssh tracing envelopes
+		// are supported. Accepting the channel indicates to clients that they may wrap their
+		// ssh payload with tracing context.
+		case tracessh.TracingChannel:
+			ch, _, err := nch.Accept()
+			if err != nil {
```
Comment (gabrielcorado) [inline lib/srv/regular/sshserver.go:1193]: Can we log this error [like in the other channel types](https://github.com/gravitational/teleport/blob/217f7240ad966b17abb9e041d4be5b9b6673998c/lib/srv/regular/sshserver.go#L1214)?

**Review by gabrielcorado — APPROVED** [review]
LGTM, just a few minor comments.

**Pull request merged**